migrations/kubernetes-agent/KubernetesAgent.UpgradeManager/KubernetesAgentUpgradeManager.cs

```diff
@@ -0,0 +1,164 @@
+        {
+            await DoBasicUpgrade(helmRelease, @namespace, newVersion, scriptExecutor, cancellationToken);
+        }
+
+        await DoUpgradeWithMigrations(helmRelease, @namespace, currentVersion, newVersion, scriptExecutor, cancellationToken);
+    }
+
+    public async Task<IVersion?> GetLatestVersion(ScriptExecutor scriptExecutor, CancellationToken cancellationToken)
+    {
+        var versionParameter = $"{latestSupportedMajorVersion.Version}.*.*{(latestSupportedMajorVersion.isPreRelease ? "-alpha.*" : string.Empty)}";
+        var getLatestVersionScript = $"helm show chart {ociRegistry} --version '{versionParameter}' 2>&1 | grep version: | awk -F'[ ]' '{{print $2}}' | tr -d '\n\t\r '";
+
+        var result = await scriptExecutor(getLatestVersionScript, [], cancellationToken);
+
+        if (result is { ExitCode: 0 } &&
+            VersionFactory.TryCreateSemanticVersion(result.Output) is { } version)
+        {
+            return version;
+        }
+
+        return null;
+    }
+
+    async Task DoUpgradeWithMigrations(string helmRelease, string @namespace, IVersion currentVersion, IVersion newVersion, ScriptExecutor scriptExecutor, CancellationToken cancellationToken)
+    {
+        var getHelmValuesScript = $"helm get values {helmRelease} --namespace={@namespace} -o json";
+        var result = await scriptExecutor(getHelmValuesScript, [], cancellationToken);
+        if (result is not { ExitCode: 0 })
+        {
+            throw new InvalidOperationException($"Failed to retrieve current helm release values with exit code {result.ExitCode}");
+        }
+
+        var valuesJson = result.Output;
+
+        var updatedValuesJson = migrationManager.MigrateValues(currentVersion, newVersion, valuesJson);
+
+        var valuesYaml = ConvertJsonToYaml(updatedValuesJson);
+
+        var upgradeWithMigrationsScript =
+            $"""
+             helm upgrade \
+             --atomic \
+             --version {newVersion} \
+             --namespace {@namespace} \
+             {helmRelease} \
+             {ociRegistry} \
+             --values values.yaml
+             """;
+
+        var upgradeResult = await scriptExecutor(upgradeWithMigrationsScript, [new File("values.yaml", Encoding.ASCII.GetBytes(valuesYaml))], cancellationToken);
+
+        if (upgradeResult is not { ExitCode: 0 })
+        {
+            throw new InvalidOperationException($"Kubernetes agent upgrade failed with exit code {upgradeResult.ExitCode}");
+        }
+    }
+
+    string ConvertJsonToYaml(string json)
+    {
+        dynamic? deserializedObject = JsonConvert.DeserializeObject<ExpandoObject>(json, new ExpandoObjectConverter());
+        if (deserializedObject is null)
+        {
+            throw new InvalidOperationException("Failed to upgrade Kubernetes agent, unable to parse json values after migration.");
+        }
+        var serializer = new YamlDotNet.Serialization.SerializerBuilder().WithQuotingNecessaryStrings(true).Build();
```

Comment:
the `.WithQuotingNecessaryStrings(true)` is the magical option to make it work as you would expect 😅